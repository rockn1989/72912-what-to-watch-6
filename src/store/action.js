export const ActionType = {
  SET_ACTIVE_GENRE: `set_genre`,
  FILTER_FILMS: `filter_films`
};

export const ActionCreator = {
  setGenre: (genre) => ({
    type: ActionType.SET_ACTIVE_GENRE,
    payload: genre
  }),
  filterFilms: () => ({
    type: ActionType.FILTER_FILMS
  })
};
