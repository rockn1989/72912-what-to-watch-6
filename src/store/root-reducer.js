import {combineReducers} from 'redux';

import {filmsData} from './films-data/films-data';
import {filmData} from './film-data/film-data';
import {user} from './user/user';
import {formStatus} from './form-status/form-status';
import {errorStatus} from './error-status/error-status';

export const NameSpace = {
  FILMS_DATA: `FILMS_DATA`,
  FILM_DATA: `FILM_DATA`,
  USER: `USER`,
  FORM_STATUS: `FORM_STATUS`,
  ERROR_STATUS: `ERROR_STATUS`
};

export default combineReducers({
  [NameSpace.FILMS_DATA]: filmsData,
  [NameSpace.FILM_DATA]: filmData,
  [NameSpace.USER]: user,
  [NameSpace.FORM_STATUS]: formStatus,
  [NameSpace.ERROR_STATUS]: errorStatus,
});

// export default combineReducers({
//   filmsData,
//   genre,
//   user,
//   formStatus,
//   errorStatus,
// });
