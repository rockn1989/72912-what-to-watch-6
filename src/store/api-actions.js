import {ActionCreator} from "./action";

export const loadFilmsList = () => (dispatch, _getState, api) => {
  api.get(`/films`)
    .then(({data}) => {
      dispatch(ActionCreator.loadFilms(data));
      dispatch(ActionCreator.filterFilms());
    });
};

export const loadFilm = (id) => (dispatch, _getState, api) => {
  api.get(`/films/${id}`)
    .then(({data}) => {
      dispatch(ActionCreator.loadFilm(data));
    })
    .catch(() => {
      dispatch(ActionCreator.redirect(`/page-not-found`));
    });
};

export const login = () => (dispatch, _getState, api) => {
  api.get(`/login`)
    .then(({data, status}) => {
      if (status !== 401) {
        dispatch(ActionCreator.authorization(true));
        dispatch(ActionCreator.setUserInfo(data));
      }
    });
};

export const logout = () => (dispatch, _getState, api) => {
  api.get(`/logout`)
    .then(({status}) => {
      if (status === 200) {
        dispatch(ActionCreator.authorization(false));
      }
    });
};

export const sendLogin = ({email, password}) => (dispatch, _getState, api) => {
  api.post(`/login`, {
    email,
    password
  }).then(({data, status}) => {
    if (status === 200) {
      dispatch(ActionCreator.authorization(true));
      dispatch(ActionCreator.setUserInfo(data));
      dispatch(ActionCreator.redirect(`/`));
    } else {
      dispatch(ActionCreator.authorization(false));
    }
  });
};

export const sendComment = ({id, comment, rating}) => (dispatch, _getState, api) => {
  dispatch(ActionCreator.sendFormData(false));
  api.post(`/comments/${id}`, {
    rating,
    comment
  }).then(() => {
    dispatch(ActionCreator.sendFormData(true));
    dispatch(ActionCreator.redirect(`/films/${id}`));
  })
  .catch(() => {
    dispatch(ActionCreator.showError(true));
  });
};
