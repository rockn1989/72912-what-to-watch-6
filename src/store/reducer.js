import {ActionType} from './action';
import {AUTH, NO_AUTH} from '../const';

const initialState = {
  genre: `All genres`,
  film: {},
  films: [],
  filtredFilmsList: [],
  filmsCounter: [],
  authorizationStatus: NO_AUTH,
  userInfo: {
    avatarUrl: `img/avatar.jpg`
  }
};

const reducer = (state = initialState, action) => {

  switch (action.type) {
    case ActionType.LOAD_FILMS:
      return {
        ...state,
        films: action.payload
      };

    case ActionType.LOAD_FILM:
      return {
        ...state,
        film: action.payload
      };

    case ActionType.SET_ACTIVE_GENRE:
      return {
        ...state,
        genre: action.payload
      };

    case ActionType.FILTER_FILMS:
      return {
        ...state,
        filtredFilmsList: state.genre === initialState.genre ? [...state.films] : state.films.filter((film) => film.genre === state.genre)
      };

    case ActionType.SHOW_MORE:
      const genreCounter = state.filmsCounter.filter(({name}) => name === state.genre);
      const idx = state.filmsCounter.indexOf(...genreCounter);

      const newEl = {
        name: state.genre,
        counter: genreCounter[0].counter + 1
      };

      const newFilmsCounter = [...state.filmsCounter.slice(0, idx), newEl, ...state.filmsCounter.slice(idx + 1)];
      return {
        ...state,
        filmsCounter: [...newFilmsCounter]
      };

    case ActionType.SET_FILMS_COUNTER:
      const counters = [];
      for (let name of action.payload) {
        counters.push({
          name,
          counter: 0
        });
      }

      return {
        ...state,
        filmsCounter: [...counters]
      };

    case ActionType.RESET_FILMS_COUNTER:
      const resetFilmsCounter = state.filmsCounter.map((counter) => {
        return {
          name: counter.name,
          counter: 0
        };
      });
      return {
        ...state,
        filmsCounter: [...resetFilmsCounter]
      };

    case ActionType.AUTHORIZED:
      return {
        ...state,
        authorizationStatus: action.payload ? AUTH : NO_AUTH
      };

    case ActionType.SET_USER_INFO:
      const {avatar_url: avatar, ...restData} = {...action.payload};
      return {
        ...state,
        userInfo: {...restData, avatarUrl: avatar}
      };
  }
  return state;
};

export {reducer};
