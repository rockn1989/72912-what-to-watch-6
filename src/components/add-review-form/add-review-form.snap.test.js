import React from 'react';
import {render, screen} from '@testing-library/react';
import {Route, Router} from 'react-router-dom';
import configureStore from 'redux-mock-store';
import * as redux from 'react-redux';
import {createMemoryHistory} from 'history';
import {NameSpace} from '../../store/root-reducer';
import AddReviewForm from './add-review-form';

const mockStore = configureStore({});

describe(`Addreview form`, () => {

  it(`AddReviewForm should render correctly`, () => {
    const store = mockStore({
      [NameSpace.USER]: {
        authorizationStatus: false,
        userInfo: {
          avatarUrl: `img/avatar.jpg`
        },
      },
      [NameSpace.FORM_STATUS]: {
        formStatus: true
      },
      [NameSpace.ERROR_STATUS]: {
        error: false
      }
    });

    const history = createMemoryHistory();
    history.push(`/films/2/review`);

    const {container} = render(
        <redux.Provider store={store}>
          <Router history={history}>
            <Route path={`/films/2/review`}>
              <AddReviewForm
                onSendUserCommentHandler={jest.fn()}
                id={2}
                formStatus
                error
              />
            </Route>
          </Router>
        </redux.Provider>

    );

    expect(container).toMatchSnapshot();
  });

  it(`AddReviewForm must show error`, () => {
    const store = mockStore({
      [NameSpace.USER]: {
        authorizationStatus: false,
        userInfo: {
          avatarUrl: `img/avatar.jpg`
        },
      },
      [NameSpace.FORM_STATUS]: {
        formStatus: true
      },
      [NameSpace.ERROR_STATUS]: {
        error: false
      }
    });

    const history = createMemoryHistory();
    history.push(`/films/2/review`);

    render(
        <redux.Provider store={store}>
          <Router history={history}>
            <Route path={`/films/2/review`}>
              <AddReviewForm
                onSendUserCommentHandler={jest.fn()}
                id={2}
                formStatus
                error={true}
              />
            </Route>
          </Router>
        </redux.Provider>

    );

    expect(screen.getByText(/Что-то пошло не так. Попробуйте отправить сообщение позже/i)).toBeInTheDocument();
  });

});

