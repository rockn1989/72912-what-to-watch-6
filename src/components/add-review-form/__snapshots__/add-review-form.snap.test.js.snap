// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Addreview form AddReviewForm should render correctly 1`] = `
<div>
  <div
    class="add-review"
  >
    <form
      action="#"
      class="add-review__form"
    >
      <div
        class="rating"
      >
        <div
          class="rating__stars"
        >
          <input
            class="rating__input"
            id="star-1"
            name="rating"
            type="radio"
            value="1"
          />
          <label
            class="rating__label"
            for="star-1"
          >
            Rating 
            1
          </label>
          <input
            class="rating__input"
            id="star-2"
            name="rating"
            type="radio"
            value="2"
          />
          <label
            class="rating__label"
            for="star-2"
          >
            Rating 
            2
          </label>
          <input
            class="rating__input"
            id="star-3"
            name="rating"
            type="radio"
            value="3"
          />
          <label
            class="rating__label"
            for="star-3"
          >
            Rating 
            3
          </label>
          <input
            class="rating__input"
            id="star-4"
            name="rating"
            type="radio"
            value="4"
          />
          <label
            class="rating__label"
            for="star-4"
          >
            Rating 
            4
          </label>
          <input
            class="rating__input"
            id="star-5"
            name="rating"
            type="radio"
            value="5"
          />
          <label
            class="rating__label"
            for="star-5"
          >
            Rating 
            5
          </label>
          <input
            class="rating__input"
            id="star-6"
            name="rating"
            type="radio"
            value="6"
          />
          <label
            class="rating__label"
            for="star-6"
          >
            Rating 
            6
          </label>
          <input
            class="rating__input"
            id="star-7"
            name="rating"
            type="radio"
            value="7"
          />
          <label
            class="rating__label"
            for="star-7"
          >
            Rating 
            7
          </label>
          <input
            class="rating__input"
            id="star-8"
            name="rating"
            type="radio"
            value="8"
          />
          <label
            class="rating__label"
            for="star-8"
          >
            Rating 
            8
          </label>
          <input
            class="rating__input"
            id="star-9"
            name="rating"
            type="radio"
            value="9"
          />
          <label
            class="rating__label"
            for="star-9"
          >
            Rating 
            9
          </label>
          <input
            class="rating__input"
            id="star-10"
            name="rating"
            type="radio"
            value="10"
          />
          <label
            class="rating__label"
            for="star-10"
          >
            Rating 
            10
          </label>
        </div>
      </div>
      <div
        class="add-review__text"
      >
        <textarea
          class="add-review__textarea"
          id="review-text"
          name="review-text"
          placeholder="Review text"
        />
        <div
          class="add-review__submit"
        >
          <button
            class="add-review__btn"
            disabled=""
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
    <div
      style="margin-top: 10px; font-size: 14px; line-height: 16px; color: red;"
    >
      Что-то пошло не так. Попробуйте отправить сообщение позже
    </div>
  </div>
</div>
`;
